### DISCUSSION TITLE
N+1 Query problem on Django queries

### DISCUSSION THREAD

**Discussion by na0495** [django-tenants/django-tenants#1114] [Q&A]
Hello everyone,

I’m encountering a performance issue in my Django tenant-based setup where I see a lot of N+1 queries in my logs. Additionally, my logs show the database repeatedly executing SET search_path = 'tenant1','public'. This is coming from the, and it seems to happen too frequently, potentially affecting application performance.

Has anyone experienced a similar problem with django-tenants ? How did you resolve excessive N+1 queries and limit the repeated calls to SET search_path? I’m specifically looking for best practices or code samples (using select_related, prefetch_related, or tenant context managers) that can help optimize the queries and reduce the overhead of repeated schema switching.

Any advice or shared experiences would be greatly appreciated. Thank you!

**Comment by lip3l** [comment]
1.N+1 queries: they usually come from accessing related objects per-row (often in DRF serializers / SerializerMethodField). Fix by adding the needed select_related (FK/OneToOne) and prefetch_related (M2M/reverse FK) on the view/queryset used by the endpoint, so the serializer doesn’t trigger extra queries per object.

2.Repeated SET search_path: django-tenants sets search_path inside the DB wrapper cursor hook. For performance, enable TENANT_LIMIT_SET_CALLS = True so it sets the search path only once per request. Also make sure you’re not switching schema/tenant manually inside loops within the same request.
If this answers your question, please mark as Accepted Answer
